content/blog/marginal-performance-improvements.md

```diff
@@ -0,0 +1,76 @@
+
+```
+a:            |-----------------|
+
+b:  |------------|              |---------|
+```
+
+should update container `a` to look like this:
+
+```
+a:  |-------------------------------------|
+```
+
+While this is *essentially* a basic algorithm problem, we found that the competing goals of time complexity, space complexity, and code complexity made things a bit more interesting. These functions power Pilosa's core query engine, so keeping them performant and bug-free is a top priority. This partly explains why not all operation-type-type combinations currently have an in-place version; if they did, it would more than double the number of operation functions.
+
+## The Solution
+
+The new `unionRunRunInPlace` went through several iterations before it settled into something both performant and maintainable. 
+
+The first attempt was the naive one:
+
+- Join the two slices of intervals into one.
+- Sort the joined slice.
+- Iterate through intervals in joined slices and merge overlapping or adjacent intervals.
+
+This resulted in simple, understandable code, which accomplished the goal of avoiding the container type conversion. Adding a sorting step felt like the wrong approach; surely we can avoid all that additional time and memory usage. If nothing else, it became a useful baseline for benchmarking.
+
+The next attempt looked great in terms of performance. The idea was to iterate over the runs in both containers in parallel, selecting two near-by runs, and then handling each possible case of overlap in a separate branch, updating the values in the first container along the way.
+
+Unfortunately, it was **so** high-performance that it was nearly unreadable. It sat in limbo for a week while we fought over the responsibility of reviewing it (one reviewer remarking "I'm not sure if I should be excited or scared by the use of the `goto`"). I'm in favor of commenting code, but when understanding code crucially depends on dense comments, it may be a sign of fragility, and future maintenance problems.
+
+Fortunately, [Kuba](https://github.com/kuba--) stepped in with a much-improved third iteration. In the interest of keeping this post relatively short, we'll just outline the approach here. Rather than iterating over *intervals*, iteration is done over *values*, while tracking a state variable indicating whether the current value is inside any interval or not. The shift to a state-based solution simplifies the code, in part by consolidating some logic into a state-update function. The inner loop again requires several branches for different interval-overlap cases, but now each branch calls the state-update function with minor differences. The resulting code is quite a bit more readable. [See for yourself](https://github.com/pilosa/pilosa/pull/2119/files)!
```

Comment:
```suggestion
Fortunately, [Kuba](https://github.com/kuba--) stepped in with a much-improved third iteration. In the interest of keeping this post relatively short, we'll just outline the approach here. Rather than iterating over *intervals*, iteration is done over *values*, while tracking a state variable indicating whether the current value is inside any interval or not. The shift to a state-based solution simplifies the code, in part by consolidating some logic into a state-update function. The inner loop again requires several branches for different interval-overlap cases, but now each branch calls the state-update function with minor differences. The resulting code is quite a bit more readable. [See for yourself](https://github.com/pilosa/pilosa/blob/55353a256773172284ed58b01c5e3ef49c4b4323/roaring/roaring.go#L3943)!
```